### PR TITLE
Use task for applying flags to all engines

### PR DESCRIPTION
**Pull request by asarium**
For me the "Apply to all" button causes the main window to become
unresponsive since I have a lot of engine versions installed. Knossos
retrieves the flags in the main thread so until all engine versions have
been checked, the main window becomes unresponsive.

This fixes that by using a background task for retrieving the engine
flags which minimizes the work that is done on the main thread.

This currently does not display a progress bar in the GUI since I don't
know how to enable that for my task. How can I enable that?

### CODE REVIEW

File: knossos/tasks.py [hunk 1]
```diff
@@ -1834,6 +1834,65 @@ def work(self, vp_file):
         self.post(vp_file)
 
 
+class ApplyEngineFlagsTaks(progress.Task):
+    def __init__(self, mods, flags, custom_flags):
+        super(ApplyEngineFlagsTaks, self).__init__()
+
+        self.custom_flags = custom_flags
+        self.mods = mods
```
Comment (ngld) [inline knossos/tasks.py]: `self.mods` tells Knossos to display the progress bar on these mods. If you want a progress bar on the bottom, you need to leave it empty.

File: knossos/tasks.py [hunk 2]
```diff
@@ -1834,6 +1834,65 @@ def work(self, vp_file):
...
+        self.flags = flags
+
+        self.title = 'Applying engine flags...'
+        self.done.connect(self.finish)
+
+        for mod in mods:
+            self.add_work((mod,))
```
Comment (ngld) [inline knossos/tasks.py]: `self.add_work` accepts a list of items. Just pass `mods` to it.

File: knossos/tasks.py [hunk 3]
```diff
@@ -1834,6 +1834,65 @@ def work(self, vp_file):
...
+
+    def work(self, mod):
+        key = '%s#%s' % (mod.mid, mod.version)
+        flag_info = None
+
+        try:
+            progress.start_task(0.0, 1.0, 'Applying to {} {}'.format(mod.mid, mod.version))
```
Comment (ngld) [inline knossos/tasks.py]: `start_task` / `finish_task` are used for nesting. To actually update the progress bar, you need to call `progress.update`. The progress message wouldn't be visible in this case since the progress bar only shows the task's title and no detailed view for this type of task has been implemented.

You don't actually need the progress functions since you don't update the progress within one work item and progress based on the amount of processed items is automatically handled by `Task`.
TL;DR: Remove the `progress.start_task` and `progress.finish_task` calls.

File: knossos/tasks.py [hunk 4]
```diff
@@ -1834,6 +1834,65 @@ def work(self, vp_file):
         self.post(vp_file)
 
 
+class ApplyEngineFlagsTaks(progress.Task):
```
Comment (ngld) [inline knossos/tasks.py]: Typo

**Review by ngld — CHANGES_REQUESTED** [review]

File: knossos/tasks.py (hunk 1 above)
Comment (asarium) [reply knossos/tasks.py]: Ah, good to know. I didn't even need that attribute and after I removed it the progress bar shows up.

One thing I noticed was that the progress bar always seems to show 0% even for other tasks. Is that intentional or some kind of bug?

File: knossos/tasks.py (hunk 2 above)
Comment (asarium) [reply knossos/tasks.py]: Ah, I thought i needed tuples but this is a lot cleaner.

File: knossos/tasks.py (hunk 3 above)
Comment (asarium) [reply knossos/tasks.py]: Yes, I was a little confused about how the progress system works. I removed all usages of those functions from the task code now.

File: knossos/tasks.py (hunk 4 above)
Comment (asarium) [reply knossos/tasks.py]: Fixed.

**Comment by ngld** [comment]
The progress bar should be working. Does it move if you check the file integrity of a mod with a bunch of packages?
You can also add `progress.update(0, '')` at the start of your `work` method and see if that helps.

**Comment by asarium** [comment]
So, checking the file integrity does show the progress bar properly but `progress.update(0, '')` doesn't seem to do anything. However, if I add the progress update and set `_threads` to 1 it works correctly. Without the `progress.update(0, '')` the progress bar still stays at 0%.

Final try: I added a random sleep time to each work call and now the progress bar also works "correctly". Maybe the tasks are done before the progress bar is updated? But then it still shows up for a short time and stays at 0%...

Anyway, it seems to be working when I add the `progress.update(0, '')` so this should be ready now.
